File: retalert/core/alert.py
"""Alert model + delivery states.

An Alert is one outgoing emergency event, fanned out to one or more
recipients. Per-recipient delivery/ack state lives in AckTracker; the Alert
itself is the payload + recipient list + retry policy. See PROMPT.md
§ Acknowledgement protocol and § Transport intelligence & failover.
"""
from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field, asdict
from typing import Dict, List, Optional


# Per-recipient delivery states (ordered severity).
SENT = "sent"            # handed to transport, no confirmation yet
DELIVERED = "delivered"  # transport confirmed delivery (LXMF DELIVERED)
ACKED = "acked"          # app-level ack from recipient (build step 9)
REPLIED = "replied"      # recipient sent a canned reply (build step 9)
FAILED = "failed"        # transport gave up / no path

# Step 2: DELIVERED is terminal (transport confirmed). Build step 9 wires
# app-level ACK and will move DELIVERED back to ACTIVE so alerts retry until
# the recipient actually acknowledges.
ACTIVE_STATES = {SENT}
DONE_STATES = {DELIVERED, ACKED, REPLIED, FAILED}

SEVERITIES = ("help", "medical", "danger", "critical")


@dataclass
class Alert:
    alert_id: str = ""
    severity: str = "help"
    text: str = ""
    recipients: List[str] = field(default_factory=list)  # dest hash hex
    created_at: float = 0.0
    # payload toggles (filled by PresetResolver in build step 10)
    payload: Dict = field(default_factory=dict)
    # retry policy
    retry_interval: float = 3.0   # seconds between attempts while unacked
    max_attempts: int = 0         # 0 = unlimited (until ack/failed)
    # transport plan: "off" | "critical" (default) | "all" (Hail Mary fan-out)
    fan_out: str = "critical"

    def __post_init__(self):
        if not self.alert_id:
            self.alert_id = uuid.uuid4().hex
        if self.created_at == 0.0:
            self.created_at = time.time()
        if self.severity not in SEVERITIES:
            self.severity = "help"

    def as_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict) -> "Alert":
        return cls(
            alert_id=d.get("alert_id") or uuid.uuid4().hex,
            severity=d.get("severity", "help"),
            text=d.get("text", ""),
            recipients=list(d.get("recipients", [])),
            created_at=float(d.get("created_at", 0.0)),
            payload=dict(d.get("payload", {})),
            retry_interval=float(d.get("retry_interval", 3.0)),
            max_attempts=int(d.get("max_attempts", 0)),
            fan_out=d.get("fan_out", "critical"),
        )

File: android/domain/src/main/kotlin/network/retalert/domain/Alert.kt
package network.retalert.domain

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Per-recipient delivery states (ordered severity). Mirrors retalert/core/alert.py. */
object AckState {
    const val SENT = "sent"          // handed to transport, no confirmation yet
    const val DELIVERED = "delivered" // transport confirmed delivery (LXMF DELIVERED)
    const val ACKED = "acked"         // app-level ack from recipient
    const val REPLIED = "replied"    // recipient sent a canned reply
    const val FAILED = "failed"      // transport gave up / no path

    val ACTIVE = setOf(SENT)
    val DONE = setOf(DELIVERED, ACKED, REPLIED, FAILED)

    val ALL = listOf(SENT, DELIVERED, ACKED, REPLIED, FAILED)
}

object Severity {
    const val HELP = "help"
    const val MEDICAL = "medical"
    const val DANGER = "danger"
    const val CRITICAL = "critical"
    val ALL = listOf(HELP, MEDICAL, DANGER, CRITICAL)
    fun normalize(s: String?): String =
        if (s != null && s in ALL) s else HELP
}

object FanOut {
    const val OFF = "off"
    const val CRITICAL = "critical"   // default
    const val ALL = "all"
    fun normalize(s: String?): String =
        if (s == OFF || s == ALL) s!! else CRITICAL
}

/** One outgoing emergency event, fanned out to one or more recipients.
 *  Mirrors `Alert` in retalert/core/alert.py. */
@Serializable
data class Alert(
    val alertId: String,
    val severity: String = Severity.HELP,
    val text: String = "",
    val recipients: List<String> = emptyList(),   // destination hash hex
    val createdAt: Double = 0.0,
    val payload: Map<String, Boolean> = emptyMap(),
    val retryInterval: Double = 3.0,
    val maxAttempts: Int = 0,                     // 0 = unlimited
    val fanOut: String = FanOut.CRITICAL,
) {
    companion object {
        /** Create with auto alert_id + createdAt (Python __post_init__ parity). */
        fun new(
            clock: Clock,
            severity: String = Severity.HELP,
            text: String = "",
            recipients: List<String> = emptyList(),
            payload: Map<String, Boolean> = emptyMap(),
            retryInterval: Double = 3.0,
            maxAttempts: Int = 0,
            fanOut: String = FanOut.CRITICAL,
        ): Alert = Alert(
            alertId = randomHex(16),
            severity = Severity.normalize(severity),
            text = text,
            recipients = recipients,
            createdAt = if (clock.nowEpoch() == 0.0) RealClock.nowEpoch() else clock.nowEpoch(),
            payload = payload,
            retryInterval = retryInterval,
            maxAttempts = maxAttempts,
            fanOut = FanOut.normalize(fanOut),
        )
    }

    fun asJsonObject(): JsonObject = buildJsonObject {
        put("alert_id", alertId)
        put("severity", severity)
        put("text", text)
        put("recipients", JsonPrimitive(recipients.joinToString(","))) // legacy shape handled in data layer
        put("created_at", createdAt)
        put("retry_interval", retryInterval)
        put("max_attempts", maxAttempts)
        put("fan_out", fanOut)
    }
}